### ISSUE TITLE
libstreflop-dev now available for testing on Ubuntu

### ISSUE THREAD

**Issue by kiplingw** [MegaGlest/megaglest-source#119]
Hey everyone,

I'm a long time fan of MegaGlest (and before that Glest). 

I recently required a library that provided reliable and portable floating point calculations and recalled streflop. I had actually discovered streflop some years prior via Glest, but until now had now had no use for it.

Like most developers, I would prefer to rely on a properly packaged library rather than having to include an embedded copy. For some operating systems, if the library is already packaged, this can violate a community policy, such as DPM 4.13:

```
https://www.debian.org/doc/debian-policy/ch-source.html#s-embeddedfiles
https://wiki.debian.org/EmbeddedCodeCopies
```

Of course that's not helpful if the library hasn't been packaged. Several other projects also use streflop, such as Spring RTS by using an embedded copy.

```
https://springrts.com/mantis/view.php?id=5058
```

I am happy to share with you my PPA for streflop which is currently under testing. Please consider testing this library with your build environment in the usual way via pkg-config. With enough feedback and successful usage, I will have it become an official Debian package.

```
https://launchpad.net/~kip/+archive/ubuntu/streflop
```

To test, please run the following on Ubuntu (wily only at this time):
$ sudo add-apt-repository ppa:kip/streflop
$ sudo apt-get update
$ sudo apt-get install libstreflop-dev

What I suggest you do is test to see if the library is already available on the user's system via pkg-config during build time, and if so, use the system installation. If not, then revert back to building your embedded version.

It was a major pain to Debianize upstream's source. Their build environment was quite broken.

**Comment by tomreyn** [comment]
Kip, thanks for working on debianizing streflop. While Filux had reviewed this issue (you posted) back in July, I think it would be nice to respond more verbosely to your motivated report. 

Indeed it would be nice to convert any embedded code into library references. However, to date, streflop has not made it to either Debian or Ubuntu (nor some of the other major Linux distributions). I acknowledge the fact that this is kind of a hen and egg problem. But our motivation to drive a process to integrate streflop into Debian is limited, since we are actually trying to move away from it (due to the bulky dependency and added complexity it introduces).

However, we are not just there, yet, so for now we _would_ benefit from it hitting major Linux distributions' repositories, and would appreciate your work in this direction.

(Please note that I am not a C++ developer and I am merely trying to sum up what I think Softcoder discussed regarding streflop in MegaGlest roughly a year ago.)

**Comment by kiplingw** [comment]
Hey tomreyn. Thanks for sharing your thoughts.

Yes, to date streflop is not an official Debian package (and consequently not in Ubuntu main / universe). I honestly haven't spent much time searching for a downstream sponsor with Debian either. Since I Debianized streflop the package has been working fine for my project's needs, so I left it at that.

Having said that, if you're able to test integrating with MegaGlest, it's easier to build the case for the sponsor later.

Also, regarding trying to move away from streflop, I am in agreement with you. Upstream is not maintaining it anymore and I had to hack quite a few patches to get it to build. In fact, it was so broken it actually managed to [crash](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=69715) GCC! 

Unfortunately at the time I didn't know of any other libraries out there that can handle floating point portability. I wasn't able to use [MPFR](http://www.mpfr.org/) (C API) due to licensing restrictions on one of the elegant C++ bindings for use in a non-free project. But I think it would probably be a much better candidate for MegaGlest moving forward. It's also properly Debianized _and_ with pkg-config support I [patched](https://lists.gforge.inria.fr/pipermail/mpfr-commits/2016-April/009843.html) in.

**Comment by softcoder** [comment]
I'll give this a try today and see what can be done

**Comment by softcoder** [comment]
This is now supported in our cmake system (I tested using your repo mentioned above)

**Comment by softcoder** [comment]
I asked the guys in debian games and they asked me to:

point the streflop packager at https://mentors.debian.net/intro-maintainers

to get this into Debian (and then it will go into Ubuntu)

**Comment by kiplingw** [comment]
Hey Mark. That's great news! Hopefully this eases the maintenance nightmare that goes with having to embed it in your local source tree.

Regarding Debian, believe me, I would love to have it merged downstream. I actually tried a while back but I had a hard time finding a sponsor. If you happen to find one, that would be great. Part of the problem is upstream's build environment was so mangled and horribly cobbled together that even with all of my Debian maintainer scripts and quilt patches, it's hard to make the sale to the sponsor.

**Comment by softcoder** [comment]
Pabs (Paul Wise) on the Debian Games team will sponser, please check that link and he will help out.

Thanks

**Comment by kiplingw** [comment]
Hey Mark. Actually Paul and I have already crossed paths several times. Regarding streflop, last April.  Paul had a number of concerns regarding upstream's build environment. Unfortunately upstream is MIA. I've reached out to the official maintainer several times with null response. 

The way I see it, the package needs a sponsor who is willing to accept it for upstream's flaws. The idea being that since streflop has no official maintainer anymore, we have to assume it is abandonware for the time being. Packaged as best we can it is more useful for other projects that still have it as a dependency than it is for them to have to embed it. Whereas new projects should probably look at MPFR as an alternative that is actively maintained.
